[serve] Integrate and Document Bring-Your-Own Gradio Applications (#26403)
ray-project/ray: doc/source/serve/tutorials/gradio-integration.md

# Scaling your Gradio app with Ray Serve

> In this guide, we will show you how to scale up your Gradio application using Ray Serve. There is no need to change the internal architecture of your Gradio app; instead, we will neatly wrap it with Ray Serve and then scale it up to access more resources.

Reviewer: Nit:

```suggestion
In this guide, we will show you how to scale up your [Gradio](https://gradio.app/) application using Ray Serve. There is no need to change the internal architecture of your Gradio app; instead, we will neatly wrap it with Ray Serve and then scale it up to access more resources.
```

## Dependencies

To follow this tutorial, you will need Ray Serve and Gradio. If you haven't already, install them by running:
```console
$ pip install "ray[serve]"
$ pip install gradio
```
For the purposes of this tutorial, we will be working with Gradio apps that run text summarization and text generation models. **Note that you can substitute this Gradio app for any Gradio app of your own!**

> We will be using HuggingFace's pipeline to access the model. First, let's install the transformers module.

Reviewer: Nit:

```suggestion
We will be using [HuggingFace's Pipelines](https://huggingface.co/docs/transformers/main_classes/pipelines) to access the model. First, let's install the transformers module.
```

```console
$ pip install transformers
```

## Basic Example: Deploy your Gradio app with Ray Serve

This example will show you an easy, straightforward way to deploy your app onto Ray Serve. Start by creating a new Python file named `demo.py` and import `GradioServer` from Ray Serve for deploying your Gradio app, `gradio`, and `transformers.pipeline` for loading text summarization models.
```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-integration.py
:start-after: __doc_import_begin__
:end-before: __doc_import_end__
```

Then, we construct the (optional) Gradio app `io`:

> > **Note**: Remember you can substitute this with your own Gradio app if you want to try scaling up your own Gradio app!

Reviewer: For all the Notes, can we maintain consistency with the other notes in the docs using this syntax:

```
:::{note}
Remember you can substitute this with your own Gradio app if you want to try scaling up your own Gradio app!
:::
```

```console
$ serve run demo:app
```

Now you can access your Gradio app at `http://localhost:8000`! This is what it should look like:
![Gradio Result](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/gradio_result.png)

See [Putting Ray Serve Deployment Graphs in Production](https://docs.ray.io/en/master/serve/production.html#id1) for more information on how to deploy your app in production.

## Complex Example: Parallelism
You can run multiple models in parallel with Ray Serve.

### Original Approach
Suppose you want to run the following program.

1. Take two text generation models, [`gpt2`](https://huggingface.co/gpt2) and [`EleutherAI/gpt-neo-125M`](https://huggingface.co/EleutherAI/gpt-neo-125M).
2. Run the two models on the same input text, such that the generated text has a minimum length of 20 and maximum length of 100.
3. Display the outputs of both models using Gradio.

This is how you would do it normally:

```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-original.py
:start-after: __doc_code_begin__
:end-before: __doc_code_end__
```

### Parallelize using Ray Serve

> With Ray Serve, we can parallelize the two text generation models by wrapping each model in a separate Ray Serve [deployment](serve-key-concepts-deployment). Deployments are defined by decorating a Python class or function with `serve.deployment`, and usually wrap the models that you want to deploy on Ray Serve and handle incoming requests.

Reviewer: Nit:

```suggestion
With Ray Serve, we can parallelize the two text generation models by wrapping each model in a separate Ray Serve [deployment](serve-key-concepts-deployment). Deployments are defined by decorating a Python class or function with `@serve.deployment`, and usually wrap the models that you want to deploy on Ray Serve and handle incoming requests.
```

> First, let's import our dependencies. Note that we need to import `GradioIngress` instead of `GradioServer` like before, since we are actually building our own deployable `MyGradioServer` with more customization that allows us to run models in parallel.

Reviewer: ```suggestion
First, let's import our dependencies. Note that we need to import `GradioIngress` instead of `GradioServer` like before since we're now building a customized `MyGradioServer` that can run models in parallel.
```

```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-integration-parallel.py
:start-after: __doc_import_begin__
:end-before: __doc_import_end__
```

Then, let's wrap our `gpt2` and `EleutherAI/gpt-neo-125M` models in Serve deployments, named `TextGenerationModel`.
```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-integration-parallel.py
:start-after: __doc_models_begin__
:end-before: __doc_models_end__
```

> Then, instead of simply wrapping our Gradio app in a Gradio Server deployment, we can build our own `MyGradioServer` that reroutes the Gradio app so that it runs the `TextGenerationModel` deployments:

Reviewer: ```suggestion
Next, instead of simply wrapping our Gradio app in a Gradio Server deployment, we can build our own `MyGradioServer` that reroutes the Gradio app so that it runs the `TextGenerationModel` deployments:
```

```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-integration-parallel.py
:start-after: __doc_gradio_server_begin__
:end-before: __doc_gradio_server_end__
```

Lastly, we link everything together:
```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-integration-parallel.py
:start-after: __doc_app_begin__
:end-before: __doc_app_end__
```

> **NOTE**: This will bind your two text generation models (wrapped in Serve deployments) to `MyGradioServer._d1` and `MyGradioServer._d2`. Thus, we have built our Gradio Interface `io` such that it calls `MyGradioServer.fanout()`, which simply sends requests to your two text generation models that are deployed on Ray Serve.

> Now, you can run your app, and the two text generation models run in parallel on a Ray Cluster while your app is also scalable at the same time! Run your Gradio app:

Reviewer: ```suggestion
Now, you can run your scalable app, and the two text generation models will run in parallel on your Ray cluster! Run your Gradio app:
```

:start-after: __doc_gradio_ingress_begin__
:end-before: __doc_gradio_ingress_end__
```

### Deploy your Gradio Server
Replicas in a deployment are copies of your program living on Ray Serve, and more replicas means your deployment can serve more client requests. You can increase the number of replicas of your application or increase the number of CPUs and/or GPUs available to each replica.

Then, using either the example we created above, or an existing Gradio app (of type `Interface`, `Block`, `Parallel`, etc.), wrap it in your Gradio Server.

```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-integration.py
:start-after: __doc_app_begin__
:end-before: __doc_app_end__
```

Finally, deploy your Gradio Server! Run the following in your terminal:
```console
$ serve run demo:app
```

Now you can access your Gradio app at `http://localhost:8000`! This is what it should look like:
![Gradio Result](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/gradio_result.png)

See [Putting Ray Serve Deployment Graphs in Production](https://docs.ray.io/en/master/serve/production.html#id1) for more information on how to deploy your app in production.

> ## Complex Example: Parallelism

Reviewer: I wonder if there's a better way to frame these sections than "Complex Example" vs "Basic Example."

How about instead of "Basic Example: Deploy your Gradio app with Ray Serve", we say "Quickstart: Deploy your Gradio app with Ray Serve"? And instead of "Complex Example: Parallelism", we say "Parallelizing models with Ray Serve"?

Author: Sounds good to me! Thanks for the suggestion!

```bash
$ pip install "ray[serve]"
$ pip install gradio
```
For the purposes of this tutorial, we will be working with Gradio Apps that runs text summarization and text generation models. **Note that you can substitute this Gradio App for any Gradio App of your own!**

We will be using HuggingFace's pipeline to access the model. First, let's install the transformers module.
```bash
$ pip install transformers
```

## Basic Example: Deploy your Gradio App with Ray Serve

In this example, we will show you the easy, straightforward way to deploying your app onto Ray. Start by creating a new python file named `demo.py` and import all dependencies.
```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-integration.py
:start-after: __doc_import_begin__
:end-before: __doc_import_end__
```

Then, we construct the (optional) Gradio app `io`:
> **Note**: Remember you can substitute this with your own Gradio app if you want to try scaling up your own Gradio app!
```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-integration.py
:start-after: __doc_gradio_app_begin__
:end-before: __doc_gradio_app_end__
```

### Understanding `GradioServer`

> In order to deploy your Gradio App onto a Ray Cluster, you need to wrap your Gradio App in a serve deployment. `GradioIngress` acts as that wrapper. It serves your Gradio App on a Ray cluster.

Reviewer 2: +1 good idea

```python
@serve.deployment
class GradioServer:
    def __init__(self, io: gr.Blocks):
        self.app = gr.routes.App.create_app(io)

    async def __call__(self, request: starlette.requests.Request):
        sender = ASGIHTTPSender()
        await self.app(request.scope, receive=request.receive, send=sender)
        return sender.build_asgi_response()
```

### Deploy your Gradio Server
You can increase the number of replicas of your application (which can take on more client requests) or increase the number of CPUs and/or GPUs available to each replica.

Then, using either the example we created above, or an existing Gradio App (of type `Interface`, `Block`, `Parallel`, etc.), wrap it in your Gradio Server.

```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-integration.py
:start-after: __doc_app_begin__
:end-before: __doc_app_end__
```

Finally, deploy your Gradio Server!
```bash
$ serve run demo:app
```

> Now you can access your Gradio app at `http://localhost:8000`! This is what it should look like:

Reviewer: Can we link to ["Deploying Ray Serve"](https://docs.ray.io/en/releases-1.13.0/serve/deployment.html) or ["Putting Ray Serve Deployment Graphs in Production"](https://docs.ray.io/en/master/serve/production.html#id1), so users can learn how to actually deploy this in production?

![Gradio Result](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/gradio_result.png)

## Complex Example: Parallelism
If you have multiple models you want to run in parallel, one way to do so is directly through Ray Serve.

### Original Approach
Suppose you want to run the following program.

1. Take two text generation models, `gpt2` and `EleutherAI/gpt-neo-125M`.
2. Run the two models on the same input text, such that the generated text has a minimum length of 20 and maximum length of 100.
3. Display the outputs of both models using Gradio.

This is how you would do it normally:

> ```python

Author: I played around with it and there doesn't seem to be a good way to test the `launch()` statement, as it will block the python script and the test statements afterwards don't get run. Synced with @shrekris-anyscale and decided not to include for now.

)
```

### Understanding `GradioIngress`
In order to deploy your Gradio App onto a Ray Cluster, you need to wrap your Gradio App in a serve deployment. `GradioIngress` acts as that wrapper. It serves your Gradio App on a Ray cluster.
```python
class GradioIngress:
    def __init__(self, io: gr.Blocks):
        self.app = gr.routes.App.create_app(io)

    async def __call__(self, request: starlette.requests.Request):
        sender = ASGIHTTPSender()
        await self.app(request.scope, receive=request.receive, send=sender)
        return sender.build_asgi_response()
```

### Define your Gradio Server
First, we need to define a deployable Gradio Server built on top of `GradioIngress`. As this example is demonstrating its most basic use as a wrapper, we just need to decorate `GradioIngress` as a serve deployment.

```python
import gradio as gr
from ray.serve.gradio import GradioIngress

GradioServer = serve.deployment(GradioIngress)
```

### Deploy your Gradio Server
First, Ray Serve needs to run on top of a Ray cluster, so we connect to a local one.

```python
ray.init(address="auto")
```

Next, we start the Ray Serve runtime:

```python

> serve.start(detached=True)

Reviewer 3: You don't need to do either of these if you use `serve run` CLI. I'd suggest doing that for this example and link out for instructions on how to use it on a cluster. The CLI command would be `serve run file:gradio_app`

```python
@serve.deployment
class GradioServer(GradioIngress):
    def __init__(self, downstream_model_1, downstream_model_2):
        self._d1 = downstream_model_1
        self._d2 = downstream_model_2

        io = gr.Interface(self.fanout, "textbox", "textbox")
        super().__init__(io)

    def fanout(self, text):
        [result1, result2] = ray.get([self._d1.remote(text), self._d2.remote(text)])
        return f"{result1}\n------------\n{result2}"
```

Lastly, we link everything together:
```python
app = GradioServer.bind(app1, app2)
```

This is the full script:
```python
from ray import serve
import ray
import gradio as gr
from ray.serve.gradio import GradioIngress
from transformers import pipeline

@serve.deployment
class GradioServer(GradioIngress):
    def __init__(self, downstream_model_1, downstream_model_2):
        self._d1 = downstream_model_1
        self._d2 = downstream_model_2

        io = gr.Interface(self.fanout, "textbox", "textbox")
        super().__init__(io)

    def fanout(self, text):
        [result1, result2] = ray.get([self._d1.remote(text), self._d2.remote(text)])

>         return f"{result1}\n------------\n{result2}"

Reviewer 3: nit: use `async def` as best practice :)

Author: Unfortunately Gradio doesn't support async functions in gr.Interface yet :(
